src/sparse_blas/backends/mkl_common/mkl_spmv.cxx

```diff
@@ -0,0 +1,229 @@
+             A_view.type_view == matrix_descr::hermitian) {
+        // No optimize_symv currently
+        return;
+    }
+    else {
+        oneapi::mkl::sparse::optimize_gemv(queue, opA, internal_A_handle->backend_handle);
+    }
+}
+
+sycl::event spmv_optimize(sycl::queue &queue, oneapi::mkl::transpose opA, const void *alpha,
+                          oneapi::mkl::sparse::matrix_view A_view,
+                          oneapi::mkl::sparse::matrix_handle_t A_handle,
+                          oneapi::mkl::sparse::dense_vector_handle_t x_handle, const void *beta,
+                          oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+                          oneapi::mkl::sparse::spmv_alg alg,
+                          oneapi::mkl::sparse::spmv_descr_t /*spmv_descr*/, void * /*workspace*/,
+                          const std::vector<sycl::event> &dependencies) {
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    bool is_beta_host_accessible = detail::is_ptr_accessible_on_host(queue, beta);
+    check_valid_spmv(__func__, opA, A_view, A_handle, x_handle, y_handle, is_alpha_host_accessible,
+                     is_beta_host_accessible);
+    auto internal_A_handle = detail::get_internal_handle(A_handle);
+    if (internal_A_handle->all_use_buffer()) {
+        detail::throw_incompatible_container(__func__);
+    }
+    if (alg == oneapi::mkl::sparse::spmv_alg::no_optimize_alg) {
+        return detail::collapse_dependencies(queue, dependencies);
+    }
+    internal_A_handle->can_be_reset = false;
+    if (A_view.type_view == matrix_descr::triangular) {
+        return oneapi::mkl::sparse::optimize_trmv(queue, A_view.uplo_view, opA, A_view.diag_view,
+                                                  internal_A_handle->backend_handle, dependencies);
+    }
+    else if (A_view.type_view == matrix_descr::symmetric ||
+             A_view.type_view == matrix_descr::hermitian) {
+        return detail::collapse_dependencies(queue, dependencies);
+    }
+    else {
+        return oneapi::mkl::sparse::optimize_gemv(queue, opA, internal_A_handle->backend_handle,
+                                                  dependencies);
+    }
+}
+
+template <typename T>
+sycl::event internal_spmv(sycl::queue &queue, oneapi::mkl::transpose opA, const void *alpha,
+                          oneapi::mkl::sparse::matrix_view A_view,
+                          oneapi::mkl::sparse::matrix_handle_t A_handle,
+                          oneapi::mkl::sparse::dense_vector_handle_t x_handle, const void *beta,
+                          oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+                          oneapi::mkl::sparse::spmv_alg /*alg*/,
+                          oneapi::mkl::sparse::spmv_descr_t /*spmv_descr*/,
+                          const std::vector<sycl::event> &dependencies,
+                          bool is_alpha_host_accessible, bool is_beta_host_accessible) {
+    T host_alpha =
```

Comment:
Thanks to @spencerpatty who brought up an important point here. The "Notes" section of the [oneAPI Spec](https://uxlfoundation.github.io/oneAPI-spec/spec/elements/oneMKL/source/domains/spblas/operations/spmv.html) says that the `spmv_optimize()` stage is compulsory. We need a way to check if the `optimize` stage has been called, and if not, throw an exception. Same applies for `spmv_buffer_size()` API.

Reply:
Also, it seems we need to clarify in the oneAPI Spec explicitly that if either `optimize`/`buffer_size` stages are not called, then it should throw a specific exception (invalid_argument?)

Reply:
I thought about it but the issue is that we would need to check that the `*_buffer_size` and `*_optimize` functions are called with the same arguments for the check to be really useful. I think it would add too much overhead to check this. I would suggest to only clarify in the spec that not doing so is undefined behavior.
I was also thinking of storing in the descriptor whether `buffer_size` or `optimize` has been called once regardless of the arguments used but this is only a partial check. We would need to justify in the spec that we can throw an exception for some cases and keep it undefined behavior if different arguments are used. It's simpler to say not calling them is undefined behavior.

Reply:
Undefined behavior is not good enough. We can't have users getting used to expecting that not having calls to `_buffer_size` or `_optimize` APIs is fine and things will work just because oneMKL backend works right now. It is okay for us to enforce restrictions first, and later lift some of them as needed, but it's not possible for us to do it the other way round because that would break users' codes. The overhead for a few `if` checks on the host is going to be far less than the actual operation itself, even for tiny matrices. Even in the unlikely event that overhead is large, that's the cost of doing business and having the APIs that we do in the oneAPI Spec.

* Not calling the `_buffer_size()`/`_optimize()` stages is not allowed. This must be caught and an exception thrown. To do that, store flags for these in the `<api>_descr` and internally initialize them to false, setting them to true as and when the APIs are called.
* Repeating calls to any of these functions with a different `transpose` operation(s) or `alg` enum with the same `<api>_descr` is not allowed. This must be caught and an exception must be thrown. That breaks all optimizations and state of the workspace. If users need a different op/alg, they need to create a new `<api>_descr` for that with a different temporary workspace. To catch this exception, internally store the `transpose` and `alg` enums in `<api>_descr` and initialize them to invalid values not among any of the existing enums, e.g. `static_cast<oneapi::mkl::transpose>(0xFF)`. That way when any API is called the first time, you set the `transpose`/`alg` enums and they are no longer allowed to change that.
* Repeating calls with a different sparsity pattern (_including_ mathematically same but just differently ordered sparsity pattern) compared to the first call is also not allowed, but we will have to make it undefined behavior because there is no efficient way to check if the users provided us a new matrix with the same sparsity pattern in the same order or not.
* Repeating calls with different `alpha`, `beta` is always allowed, of course.

We need to clarify these things in the oneAPI specificiation.

Reply:
or we just lift the requirement and go back to it being optional :)

Reply:
> or we just lift the requirement and go back to it being optional :)

As I mentioned in my previous comment:
> It is okay for us to enforce restrictions first, and later lift some of them as needed, but it's not possible for us to do it the other way round because that would break users' codes.

Reply:
The other reason to prefer a compulsory call to `_buffer_size`, `_optimize` APIs is because even though some backends (like the current Intel oneMKL) may not need a workspace associated with a particular `op`/`alg` combo, there can be another backend that requires for a temporary workspace some or all their algorithms thereby requiring these APIs be called. When a user tries to then run the same thing on two different hardware/backends relying on the APIs being optional, they are guaranteed to run into problems. The only way to prevent that in that case is to have non-intersecting algorithm enums or backend-specific enums in the oneAPI spec itself, which is likely not what we want.

Reply:
Another issue that will need to be undefined behavior is to check whether the `_buffer_size` function has already been called before `_optimize` with the same arguments. Nothing prevents the user from calling `_buffer_size` multiple times in a row with different arguments (but the same descriptor) or from re-using the output of `_buffer_size` from a previous call. We would need to store a map of all the set of arguments that `_buffer_size` has been called with which will introduce too much overhead. We could check that `buffer_size` has been called at least once but specifying this in the specification will look weird.
I will see if we can add some checks as you suggested but I suspect this will lead to more discussions and delay. Remember we can always improve the spec later if we find this is an issue for users.

Reply:
I have added some checks in https://github.com/oneapi-src/oneMKL/pull/500/commits/5f8e1838b612370e02e4dda55e43d938be731c6c and the associated spec change is in https://github.com/uxlfoundation/oneAPI-spec/pull/542/commits/fda9f80c8c37d47afbea48cc2db8e9df0b6a6024.
I don't think I can add more checks that wouldn't require storing a map of the arguments used.

Reply:
Some minor changes requested to the implemented checks. https://github.com/uxlfoundation/oneAPI-spec/commit/fda9f80c8c37d47afbea48cc2db8e9df0b6a6024 looks fine to me, please open a Spec PR when you have a chance, I'll approve it.

Reply:
The change if part of the existing spec PR: https://github.com/uxlfoundation/oneAPI-spec/pull/542

Reply:
I'm reading between the lines of what the [cusparseSpSV documentation](https://docs.nvidia.com/cuda/cusparse/#cusparsespsv) is _not_ saying below:
> all parameters must be consistent across `cusparseSpSV` API calls and the matrix descriptions and `externalBuffer` must not be modified between `cusparseSpSV_analysis()` and `cusparseSpSV_solve()`

which is that they haven't explicitly mentioned whether args can change if `cusparseSpSV_ext_bufferSize API is called or not. 

> Nothing prevents the user from calling _buffer_size multiple times in a row with different arguments (but the same descriptor) or from re-using the output of _buffer_size from a previous call.

My thought was that _we_ prevent the user from doing this by explicitly disallowing it. However, I modified and ran a [cusparseSpSV example](https://github.com/NVIDIA/CUDALibrarySamples/tree/master/cuSPARSE/spsv_csr) trying to repeat API calls and modify some of the parameters. cuSPARSE doesn't complain about it but **silently returns incorrect results if 2nd/3rd stages are mixed** which is expected, but **not if 1st stage is mixed**, which I was not expecting. I do see an advantage of users calling `buffer_size` API with all set of arguments they need once at the top, allocating once, but then when calling `<api>_optimize` or `<api>`, they are careful about not modifying anything including the workspace in between.

I am fine with the current state that you have, then.